parquet-variant/src/main/java/org/apache/parquet/variant/Variant.java

```diff
@@ -0,0 +1,546 @@
+      .appendPattern("HH:mm:ss")
+      .optionalStart()
+      .appendFraction(ChronoField.MICRO_OF_SECOND, 0, 6, true)
+      .optionalEnd()
+      .toFormatter(Locale.US);
+
+  /**
+   * The format for a timestamp without time zone, with nanosecond precision, truncating
+   * trailing nanosecond zeros.
+   */
+  private static final DateTimeFormatter TIMESTAMP_NANOS_NTZ_TRUNC_FORMATTER = new DateTimeFormatterBuilder()
+      .append(DateTimeFormatter.ISO_LOCAL_DATE)
+      .appendLiteral('T')
+      .appendPattern("HH:mm:ss")
+      .optionalStart()
+      .appendFraction(ChronoField.NANO_OF_SECOND, 0, 9, true)
+      .optionalEnd()
+      .toFormatter(Locale.US);
+
+  /** The format for a timestamp with time zone, truncating trailing microsecond zeros. */
+  private static final DateTimeFormatter TIMESTAMP_TRUNC_FORMATTER = new DateTimeFormatterBuilder()
+      .append(TIMESTAMP_NTZ_TRUNC_FORMATTER)
+      .appendOffset("+HH:MM", "+00:00")
+      .toFormatter(Locale.US);
+
+  /**
+   * The format for a timestamp with time zone, with nanosecond precision, truncating trailing
+   * nanosecond zeros.
+   */
+  private static final DateTimeFormatter TIMESTAMP_NANOS_TRUNC_FORMATTER = new DateTimeFormatterBuilder()
+      .append(TIMESTAMP_NANOS_NTZ_TRUNC_FORMATTER)
+      .appendOffset("+HH:MM", "+00:00")
+      .toFormatter(Locale.US);
+
+  /** The format for a time, truncating trailing microsecond zeros. */
+  private static final DateTimeFormatter TIME_TRUNC_FORMATTER = new DateTimeFormatterBuilder()
+      .appendPattern("HH:mm:ss")
+      .optionalStart()
+      .appendFraction(ChronoField.MICRO_OF_SECOND, 0, 6, true)
+      .optionalEnd()
+      .toFormatter(Locale.US);
+
+  private static Instant microsToInstant(long microsSinceEpoch) {
+    return Instant.EPOCH.plus(microsSinceEpoch, ChronoUnit.MICROS);
+  }
+
+  private static Instant nanosToInstant(long timestampNanos) {
+    return Instant.EPOCH.plus(timestampNanos, ChronoUnit.NANOS);
+  }
+
+  private static void toJsonImpl(
+      byte[] value, byte[] metadata, int pos, JsonGenerator gen, ZoneId zoneId, boolean truncateTrailingZeros)
+      throws IOException {
+    switch (VariantUtil.getType(value, pos)) {
+      case OBJECT:
+        VariantUtil.handleObject(value, pos, (size, idSize, offsetSize, idStart, offsetStart, dataStart) -> {
+          try {
+            gen.writeStartObject();
+            for (int i = 0; i < size; ++i) {
+              ObjectField field = getFieldAtIndex(
+                  i, value, metadata, idSize, offsetSize, idStart, offsetStart, dataStart);
+              gen.writeFieldName(field.key);
+              toJsonImpl(
+                  field.value.value,
+                  field.value.metadata,
+                  field.value.pos,
+                  gen,
+                  zoneId,
+                  truncateTrailingZeros);
+            }
+            gen.writeEndObject();
+          } catch (IOException e) {
+            throw new RuntimeException(e);
```

Comment:
Rather than adding try/catch inline, I think it makes more sense for `toJsonImpl` to throw `IOException` so this can be handled in the wrapper methods that are public.